matchzoo/preprocessor/dssm_preprocessor.py

```diff
@@ -45,98 +44,72 @@ class DSSMPreprocessor(engine.BasePreprocessor):
+        units = self._preprocess_units()
 
-        # Loop through user input to generate tri-letters.
-        # Used for build vocabulary of tri-letters (get dimension).
-        for idx, row in tqdm(self.datapack.left.iterrows()):
-            # For each piece of text, apply process unit sequentially.
-            text = row.text_left
+        def transform_and_extend_vocab(text):
             for unit in units:
                 text = unit.transform(text)
             vocab.extend(text)
 
-        for idx, row in tqdm(self.datapack.right.iterrows()):
-            # For each piece of text, apply process unit sequentially.
-            text = row.text_right
-            for unit in units:
-                text = unit.transform(text)
-            vocab.extend(text)
+        tqdm.pandas(desc="Preparing `text_left`.")
+        datapack.left['text_left'].progress_apply(transform_and_extend_vocab)
+        tqdm.pandas(desc="Preparing `text_right`.")
+        datapack.right['text_right'].progress_apply(transform_and_extend_vocab)
 
-        # Initialize a vocabulary process unit to build tri-letter vocab.
         vocab_unit = preprocessor.VocabularyUnit()
-        vocab_unit.fit(vocab)
+        vocab_unit.fit(tqdm(vocab, desc='Fitting vocabulary unit.'))
 
-        # Store the fitted parameters in context.
-        self.datapack.context['term_index'] = vocab_unit.state['term_index']
+        self._context['term_index'] = vocab_unit.state['term_index']
         dim_triletter = len(vocab_unit.state['term_index']) + 1
-        self.datapack.context['input_shapes'] = [(dim_triletter,),
-                                                 (dim_triletter,)]
+        self._context['input_shapes'] = [(dim_triletter,), (dim_triletter,)]
+
         return self
 
     @utils.validate_context
-    def transform(
-        self,
-        inputs: typing.List[tuple],
-        stage: str
-    ) -> datapack.DataPack:
+    def transform(self, datapack: DataPack) -> DataPack:
         """
         Apply transformation on data, create `tri-letter` representation.
 
-        :param inputs: Inputs to be preprocessed.
-        :param stage: Pre-processing stage, `train`, `evaluate`, `predict`.
+        :param datapack: Inputs to be preprocessed.
 
         :return: Transformed data as :class:`DataPack` object.
         """
-        if stage in ['evaluate', 'predict']:
-            self.datapack = segment(inputs, stage=stage)
+        datapack_copy = datapack.copy()
 
-        logger.info(f"Start processing input data for {stage} stage.")
+        term_index = self._context['term_index']
+        hashing_unit = preprocessor.WordHashingUnit(term_index)
+        units = self._preprocess_units() + [hashing_unit]
 
-        # do preprocessing from scrach.
-        units = self._prepare_process_unit()
-        # prepare word hashing unit.
-        units.append(
-            preprocessor.WordHashingUnit(self.datapack.context['term_index']))
+        for unit in units:
```

Comment:
Why not move the assignment out of the loop? You can pick out `text_left` and `text_right` before the loop, and call the `unit.transform` in the loop, then, assign the transformed text to the `datapack_copy`?

Reply:
It's just an alternative. I thought about that but there's no an obviously better way of doing this operation here.